### PR TITLE
[refactor] Refactor test selection phase

### LINKED ISSUE #488: Abstract test selection phase of the frontend in a separate class, so that it can be unit tested efficiently

### PR DESCRIPTION
**Pull request by ChristopherBignamini**
Closes #488

### CODE REVIEW

**Comment by vkarak** [comment]
@ChristopherBignamini Check [here](https://github.com/eth-cscs/reframe/wiki/Coding-Style-Guide) for the exact error codes we ignore. I don't know why pep8speaks reports a couple of them.

File: reframe/frontend/cli.py [hunk 1]
```diff
@@ -375,46 +375,27 @@ def main():
         # Locate and load checks
         try:
             checks_found = loader.load_all()
-        except OSError as e:
+        except OSError as oe:
```
Comment (teojgo) [inline reframe/frontend/cli.py]: Why did you change to `oe`. If the change is valid the next line should raise from `oe`.

Comment (vkarak) [reply reframe/frontend/cli.py]: The `e` should remain, unless its used elsewhere. This is a convention throughout the code. All exception objects are named `e`.

File: unittests/test_check_filters.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import unittest
+
+import reframe.frontend.check_filters as filters
+from reframe.core.pipeline import RegressionTest
+
+class TestCheckFilters(unittest.TestCase):
+    def setUp(self):
+        #TODO: use correct initialization and add checks
+        self.check = RegressionTest()
+        self.check.name = "check1"
```
Comment (teojgo) [inline unittests/test_check_filters.py]: Use single quotes

File: reframe/frontend/check_filters.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+def have_prgenv(prgenv):
+    def _fn(c):
+        if prgenv:
+            return util.allx(c.supports_environ(e) for e in prgenv)
+        else:
+            return bool(c.valid_prog_environs)
+
+    return _fn
+
+def have_system():
```
Comment (teojgo) [inline reframe/frontend/check_filters.py]: Since `have_system` does not take any arguments, why return a function and not directly the result?

File: reframe/frontend/check_filters.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    def _fn(c):
+        return any([c.supports_system(s.fullname)
+                    for s in rt.runtime().system.partitions])
+
+    return _fn
+
+def is_gpu_only():
```
Comment (teojgo) [inline reframe/frontend/check_filters.py]: Same as above.

File: reframe/frontend/check_filters.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    def _fn(c):
+        return c.num_gpus_per_node > 0
+
+    return _fn
+
+def is_cpu_only():
```
Comment (teojgo) [inline reframe/frontend/check_filters.py]: Same as above.

File: reframe/frontend/check_filters.py (hunk 4 above)
Comment (vkarak) [reply reframe/frontend/check_filters.py]: I would simply change the name of this function to `have_gpu_only()`.

File: reframe/frontend/check_filters.py (hunk 5 above)
Comment (vkarak) [reply reframe/frontend/check_filters.py]: I would simply change the name of this function to `have_cpu_only()`.

File: reframe/frontend/cli.py [hunk 6]
```diff
@@ -375,46 +375,27 @@ def main():
...
 
         if options.names:
-            checks_matched = filter(
-                lambda c: c if c.name in options.names else None,
-                checks_matched
-            )
+            checks_matched = filter(filters.have_name(options.names), checks_matched)
 
         # Filter checks by tags
         user_tags = set(options.tags)
```
Comment (vkarak) [inline reframe/frontend/cli.py]: There is no need for this one any more. The `have_tags` converts its argument to a set already.

File: reframe/frontend/cli.py [hunk 7]
```diff
@@ -375,46 +375,27 @@ def main():
...
-        checks_matched = filter(
-            lambda c: c if user_tags.issubset(c.tags) else None,
-            checks_matched
-        )
+        checks_matched = filter(filters.have_tag(options.tags), checks_matched)
 
+        # TODO: should we move this in utils?
```
Comment (vkarak) [inline reframe/frontend/cli.py]: Which one?

File: unittests/test_check_filters.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        self.check.tags = {'a', 'b', 'c', 'd'} 
+        self.check.valid_prog_environs = ['PrgEnv-cray', 'PrgEnv-gnu']
```
Comment (vkarak) [inline unittests/test_check_filters.py]: I would put more generic names, such as `env1`, `env2` etc.

File: unittests/test_check_filters.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        self.check.valid_systems = ['daint:gpu', 'daint:mc',
+            'dom:gpu', 'dom:mc', 'kesch:pn', 'generic:login']
```
Comment (vkarak) [inline unittests/test_check_filters.py]: Same here.

File: unittests/test_check_filters.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        self.check.num_gpus_per_node = 1
+
+    def test_have_name(self):
+        self.assertTrue((filters.have_name('check1'))(self.check))
```
Comment (vkarak) [inline unittests/test_check_filters.py]: I see too many parentheses here.

File: unittests/test_check_filters.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
+import unittest
+
+import reframe.frontend.check_filters as filters
+from reframe.core.pipeline import RegressionTest
+
+class TestCheckFilters(unittest.TestCase):
+    def setUp(self):
+        #TODO: use correct initialization and add checks
+        self.check = RegressionTest()
```
Comment (vkarak) [inline unittests/test_check_filters.py]: You should use more checks here.

File: unittests/test_check_filters.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        self.assertTrue((filters.have_tag(['a', 'c']))(self.check)) 
+        self.assertFalse((filters.have_tag(['a', 'b', 'c', 'd', 'e']))(self.check)) 
+        self.assertFalse((filters.have_tag(['x', 'y', 'z']))(self.check)) 
+
+    def test_have_prgenv(self):
+        self.assertTrue((filters.have_prgenv(['PrgEnv-cray', 'PrgEnv-gnu']))(self.check))
+        self.assertFalse((filters.have_prgenv(['PrgEnv-pgi']))(self.check))
+
+    def test_system(self):
+        self.assertTrue((filters.have_system())(self.check))
```
Comment (vkarak) [inline unittests/test_check_filters.py]: If `have_system()` does not take an argument you rely here on the actual system that you are running the unit tests. That's why `have_system()` must accept the system name as an argument.

File: reframe/frontend/cli.py (hunk 1 above)
Comment (ChristopherBignamini) [reply reframe/frontend/cli.py]: Sorry, that is just a typo.

File: reframe/frontend/cli.py (hunk 6 above)
Comment (ChristopherBignamini) [reply reframe/frontend/cli.py]: Removed.

File: reframe/frontend/check_filters.py (hunk 5 above)
Comment (ChristopherBignamini) [reply reframe/frontend/check_filters.py]: Done

File: reframe/frontend/check_filters.py (hunk 4 above)
Comment (ChristopherBignamini) [reply reframe/frontend/check_filters.py]: Done

File: reframe/frontend/cli.py (hunk 7 above)
Comment (ChristopherBignamini) [reply reframe/frontend/cli.py]: Comment removed, it referred to the old filtering version

File: unittests/test_check_filters.py (hunk 8 above)
Comment (ChristopherBignamini) [reply unittests/test_check_filters.py]: done

File: unittests/test_check_filters.py (hunk 9 above)
Comment (ChristopherBignamini) [reply unittests/test_check_filters.py]: done

File: reframe/frontend/check_filters.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+            return util.allx(c.supports_environ(e) for e in prgenv)
+        else:
+            return bool(c.valid_prog_environs)
+
+    return _fn
+
+def have_system(partitions):
+    def _fn(c):
+        return any([c.supports_system(s.fullname)
+                    for s in partitions])
```
Comment (vkarak) [inline reframe/frontend/check_filters.py]: Put everything in a single line here. It should fit.

File: unittests/test_check_filters.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
+import unittest
+
+import reframe.core.runtime as rt
+import reframe.frontend.check_filters as filters
+import unittests.fixtures as fixtures
+from reframe.core.systems import SystemPartition
+from reframe.core.pipeline import RegressionTest
+from reframe.utility.typecheck import Str
```
Comment (vkarak) [inline unittests/test_check_filters.py]: This import is not needed.

File: unittests/test_check_filters.py [hunk 15]
```diff
@@ -0,0 +1,65 @@
+import unittest
+
+import reframe.core.runtime as rt
+import reframe.frontend.check_filters as filters
+import unittests.fixtures as fixtures
+from reframe.core.systems import SystemPartition
```
Comment (vkarak) [inline unittests/test_check_filters.py]: This import is not needed.

File: unittests/test_check_filters.py [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+        self.check1.valid_prog_environs = ['env1', 'env2']
+        self.check1.valid_systems = ['testsys:gpu', 'testsys2:mc']
+        self.check1.num_gpus_per_node = 1
+        #check 2
+        self.check2 = RegressionTest()
+        self.check2.name = 'check2'
+        self.check2.tags = {'x', 'y', 'z'}
+        self.check2.valid_prog_environs = ['env3']
+        self.check2.valid_systems = ['testsys:mc', 'testsys2:mc']
+        self.check2.num_gpus_per_node = 0
```
Comment (vkarak) [inline unittests/test_check_filters.py]: You should not individual tests here. You should have a list of tests instead:

```python
def create_check(attrs):
    ret = RegressionTest()
    for k, v in attrs.items():
        setattr(ret, k, v)

    return ret

self.checks = [create_check({'name': 'check1', ...}), create_check({'name': 'check2', ...})]
```

All subsequent unit tests should act on the list of checks `self.checks`. See example in my comments below.

File: unittests/test_check_filters.py [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+
+    def test_have_name(self):
+        self.assertTrue(filters.have_name('check1')(self.check1))
```
Comment (vkarak) [inline unittests/test_check_filters.py]: This is the correct way to test this, because it reflects the way we are using the filter functions.

```python
import reframe.utility.sanity as sn

self.assertEqual(1, sn.count(filter(filters.have_name('check1'), self.checks)))
```

The reason that you cannot use `len()` directly on the result of the `filter()` function is because `filter()` returns an generator, so you would have to create a concrete data structure (e.g., `list`) out of it first. `sn.count` handles this case automatically, but it consumes the generator (not a problem in our case). It also returns a deferrable expression, but since it's included in boolean check, it will be automatically evaluated.

File: unittests/test_check_filters.py [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+
+    def test_have_prgenv(self):
+        self.assertTrue(filters.have_prgenv(['env1', 'env2'])(self.check1))
+        self.assertFalse(filters.have_prgenv(['env3'])(self.check1))
+        self.assertTrue(filters.have_prgenv(['env3'])(self.check2))
+        self.assertFalse(filters.have_prgenv(['env1', 'env2'])(self.check2))
+
+    @rt.switch_runtime(fixtures.TEST_SITE_CONFIG, 'testsys')
+    def test_system(self):
+        self.assertTrue(filters.have_system(rt.runtime().system.partitions)(self.check1))
```
Comment (vkarak) [inline unittests/test_check_filters.py]: Here you should explicitly look for a system partition by name. For example,

```python
have_system(['testsys'])
have_system(['testsys:gpu', 'testsys:mc'])
```

File: reframe/frontend/check_filters.py (hunk 13 above)
Comment (ChristopherBignamini) [reply reframe/frontend/check_filters.py]: Done.

File: unittests/test_check_filters.py (hunk 15 above)
Comment (ChristopherBignamini) [reply unittests/test_check_filters.py]: Done.

File: unittests/test_check_filters.py (hunk 14 above)
Comment (ChristopherBignamini) [reply unittests/test_check_filters.py]: Done.

**Comment by ChristopherBignamini** [comment]
@jenkins-cscs retry dom

**Comment by vkarak** [comment]
@teojgo Can you approve this PR?

File: unittests/test_check_filters.py [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+                                            self.checks)))
+        self.assertEqual(1, sn.count(filter(filters.have_name('check3'),
+                                            self.checks)))
+        self.assertEqual(0, sn.count(filter(filters.have_name('check4'),
+                                            self.checks)))
+
+    def test_have_not_name(self):
+        self.assertEqual(1, sn.count(filter(filters.have_not_name('check1'),
+                                            self.checks)))
+        self.assertEqual(1, sn.count(filter(filters.have_not_name('check2'),
```
Comment (teojgo) [inline unittests/test_check_filters.py]: Shouldn't this and the next one be equal to `2`?

Comment (vkarak) [reply unittests/test_check_filters.py]: Yes, I haven't run the unit tests again. I am fixing it now and will re-push.

**Comment by vkarak** [comment]
@teojgo It should be fine now.
